lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,515 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const ReportGenerator = require('../../../lighthouse-core/report/report-generator');
+// TODO: We only need getFilenamePrefix from asset-saver. Tree shake!
+const getFilenamePrefix = require('../../../lighthouse-core/lib/asset-saver.js').getFilenamePrefix;
+
+const firebase = require('firebase/app');
+require('firebase/auth');
+const idb = require('idb-keyval');
+
+// Polyfills
+// TODO: dynamically load this based on feature detection.
+require('whatwg-fetch');
+const URLSearchParams = require('url-search-params');
+
+const LH_CURRENT_VERSION = require('../../../package.json').version;
+const APP_URL = `${location.origin}${location.pathname}`;
+
+/**
+ * Logs messages via a UI butter.
+ * @class
+ */
+class Logger {
+  constructor(selector) {
+    this.el = document.querySelector(selector);
+  }
+
+  /**
+   * Shows a butter bar.
+   * @param {!string} msg The message to show.
+   * @param {boolean=} autoHide True to hide the message after a duration.
+   *     Default is true.
+   */
+  log(msg, autoHide = true) {
+    clearTimeout(this._id);
+
+    this.el.textContent = msg;
+    this.el.classList.add('show');
+    if (autoHide) {
+      this._id = setTimeout(_ => {
+        this.el.classList.remove('show');
+      }, 7000);
+    }
+  }
+
+  warn(msg) {
+    console.warn(msg);
+  }
+
+  error(msg) {
+    console.error(msg);
+  }
+
+  /**
+   * Explicitly hides the butter bar.
+   */
+  hide() {
+    clearTimeout(this._id);
+    this.el.classList.remove('show');
+  }
+}
+
+const logger = new Logger('#log');
+
+/**
+ * Wrapper for Firebase Authentication
+ * @class
+ */
+class FirebaseAuth {
+  constructor() {
+    this.accessToken = null;
+    this.user = null;
+
+    this.provider = new firebase.auth.GithubAuthProvider();
+    this.provider.addScope('gist');
+
+    firebase.initializeApp({
+      apiKey: 'AIzaSyApMz8FHTyJNqqUtA51tik5Mro8j-2qMcM',
+      authDomain: 'lighthouse-viewer.firebaseapp.com',
+      databaseURL: 'https://lighthouse-viewer.firebaseio.com',
+      storageBucket: 'lighthouse-viewer.appspot.com',
+      messagingSenderId: '962507201498'
+    });
+
+    // Wrap auth state callback in a promise so  other parts of the app that
+    // require login can hook into the changes.
+    this.ready = new Promise((resolve, reject) => {
```

Comment:
is this weird after you log in and then log out? `firebaseAuth.ready` will still resolve on `user` but there won't be an `accessToken` anymore